### PR TITLE
perf: refactor various timestamp caches into ES6 Maps

### PR DESCRIPTION
**Pull request by probablyup**
**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**

modified existing unit tests

**If relevant, link to documentation update:**

n/a

**Summary**

This change satisfies https://github.com/webpack/webpack/issues/6234 which desires to change the timestamp caches from plain objects to the ES6 Map type to avoid deopts, since v8 expects things to be added to Maps but objects are only fast if properties aren't added dynamically after initial assignment.

**Does this PR introduce a breaking change?**

I think yes if anything is downstream is using `compiler.fileTimestamps` or `compiler.contextTimestamps`.

### CODE REVIEW

File: test/NodeWatchFileSystem.unittest.js [hunk 1]
```diff
@@ -68,7 +68,7 @@ describe("NodeWatchFileSystem", function() {
 			if(err) throw err;
 			filesModified.should.be.eql([fileDirect]);
 			dirsModified.should.be.eql([]);
-			Object.assign({}, fileTimestamps).should.have.property(fileDirect).have.type("number");
+			fileTimestamps.get(fileDirect).should.have.type("number");
```
Comment (ooflorent) [inline test/NodeWatchFileSystem.unittest.js]: Change to:

```js
(typeof fileTimestamps.get(fileDirect)).should.be.eql("number");
```

The previous line would throw a `TypeError`.

File: lib/CachePlugin.js [hunk 2]
```diff
@@ -53,13 +53,14 @@ class CachePlugin {
...
-						fileTs[file] = +stat.mtime || Infinity;
+						fileTs.set(file, +stat.mtime || Infinity);
+
 						callback();
 					});
 				}, err => {
 					if(err) return callback(err);
-					Object.keys(fileTs).forEach(key => {
-						fileTs[key] += this.FS_ACCURENCY;
+					Array.from(fileTs.keys()).forEach(key => {
```
Comment (ooflorent) [inline lib/CachePlugin.js]: You don't need this:

```js
for(const [file, ts] of fileTs) {
  fileTs.set(file, ts + this.FS_ACCURENCY);
}
```

File: lib/Compiler.js [hunk 3]
```diff
@@ -177,8 +177,8 @@ class Compiler extends Tapable {
 	}
 
 	watch(watchOptions, handler) {
-		this.fileTimestamps = {};
-		this.contextTimestamps = {};
+		this.fileTimestamps = new Map();
```
Comment (ooflorent) [inline lib/Compiler.js]: Is it required? I think `.clear()` would be better.

**Comment by sokra** [comment]
Thanks

**Pull request merged**